## Project 2
Imagine you're working with Sprint, one of the biggest telecom companies in the USA. They're really keen on figuring out how many customers might decide to leave them in the coming months. Luckily, they've got a bunch of past data about when customers have left before, as well as info about who these customers are, what they've bought, and other things like that. So, if you were in charge of predicting customer churn, how would you go about using machine learning to make a good guess about which customers might leave? What steps would you take to create a machine learning model that can predict if someone's going to leave or not?

**DataSet** Telco customer churn: IBM dataset

**Source**
This data is taken from https://www.kaggle.com/datasets/yeanzc/telco-customer-churn-ibm-dataset

Customer churn, Exploratory Data Analysis and Analysis

### Exploratory Data Analysis

#### Import Libraries and Dataset




In [1]:
%pip install plotly

Note: you may need to restart the kernel to use updated packages.


In [2]:
%pip install imblearn

Note: you may need to restart the kernel to use updated packages.


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from IPython.display import Image
import branca.colormap as cm
import plotly.express as px
from plotly.offline import init_notebook_mode, iplot
from sklearn.preprocessing import LabelEncoder
import imblearn
from collections import Counter
from imblearn.over_sampling import SMOTE
from imblearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, classification_report
from sklearn.metrics import accuracy_score, f1_score,recall_score,precision_score
from sklearn.metrics import average_precision_score, roc_auc_score, roc_curve, auc
from xgboost import XGBClassifier
import matplotlib
import imblearn
import folium
from imblearn.over_sampling import SMOTE
from imblearn.pipeline import Pipeline

import warnings
warnings.filterwarnings('ignore')

ModuleNotFoundError: No module named 'pandas'

In [4]:
#jupyter nbextension enable keplergl --py --sys-prefix

In [5]:
import ipywidgets
import keplergl
import h3
import h3pandas

In [6]:
df = pd.read_excel('C:/Users/PYTHON/Data Science Projects/Customer churn rate/input/Telco_customer_churn.xlsx')

In [7]:
df.head()

CustomerID  Count        Country       State         City  Zip Code  \
0  3668-QPYBK      1  United States  California  Los Angeles     90003   
1  9237-HQITU      1  United States  California  Los Angeles     90005   
2  9305-CDSKC      1  United States  California  Los Angeles     90006   
3  7892-POOKP      1  United States  California  Los Angeles     90010   
4  0280-XJGEX      1  United States  California  Los Angeles     90015   

                 Lat Long   Latitude   Longitude  Gender  ...        Contract  \
0  33.964131, -118.272783  33.964131 -118.272783    Male  ...  Month-to-month   
1   34.059281, -118.30742  34.059281 -118.307420  Female  ...  Month-to-month   
2  34.048013, -118.293953  34.048013 -118.293953  Female  ...  Month-to-month   
3  34.062125, -118.315709  34.062125 -118.315709  Female  ...  Month-to-month   
4  34.039224, -118.266293  34.039224 -118.266293    Male  ...  Month-to-month   

  Paperless Billing             Payment Method  Monthly Charges Total Charges  \
0               Yes               Mailed check            53.85        108.15   
1               Yes           Electronic check            70.70        151.65   
2               Yes           Electronic check            99.65         820.5   
3               Yes           Electronic check           104.80       3046.05   
4               Yes  Bank transfer (automatic)           103.70        5036.3   

  Churn Label Churn Value Churn Score  CLTV                   Churn Reason  
0         Yes           1          86  3239   Competitor made better offer  
1         Yes           1          67  2701                          Moved  
2         Yes           1          86  5372                          Moved  
3         Yes           1          84  5003                          Moved  
4         Yes           1          89  5340  Competitor had better devices  

[5 rows x 33 columns]

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 33 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CustomerID         7043 non-null   object 
 1   Count              7043 non-null   int64  
 2   Country            7043 non-null   object 
 3   State              7043 non-null   object 
 4   City               7043 non-null   object 
 5   Zip Code           7043 non-null   int64  
 6   Lat Long           7043 non-null   object 
 7   Latitude           7043 non-null   float64
 8   Longitude          7043 non-null   float64
 9   Gender             7043 non-null   object 
 10  Senior Citizen     7043 non-null   object 
 11  Partner            7043 non-null   object 
 12  Dependents         7043 non-null   object 
 13  Tenure Months      7043 non-null   int64  
 14  Phone Service      7043 non-null   object 
 15  Multiple Lines     7043 non-null   object 
 16  Internet Service   7043 

## Checking for missing values

In [9]:
# change the datatype of Total charges
df['Total Charges'] = pd.to_numeric(df['Total Charges'], errors='coerce')

In [10]:
# check for missing values
df.isnull().sum()

CustomerID              0
Count                   0
Country                 0
State                   0
City                    0
Zip Code                0
Lat Long                0
Latitude                0
Longitude               0
Gender                  0
Senior Citizen          0
Partner                 0
Dependents              0
Tenure Months           0
Phone Service           0
Multiple Lines          0
Internet Service        0
Online Security         0
Online Backup           0
Device Protection       0
Tech Support            0
Streaming TV            0
Streaming Movies        0
Contract                0
Paperless Billing       0
Payment Method          0
Monthly Charges         0
Total Charges          11
Churn Label             0
Churn Value             0
Churn Score             0
CLTV                    0
Churn Reason         5174
dtype: int64

The Churn Reason has many missing values

In [11]:
df.groupby('Churn Label')['CustomerID'].nunique()

Churn Label
No     5174
Yes    1869
Name: CustomerID, dtype: int64

In [12]:
df[df['Total Charges'].isna()]

CustomerID  Count        Country       State            City  Zip Code  \
2234  4472-LVYGI      1  United States  California  San Bernardino     92408   
2438  3115-CZMZD      1  United States  California    Independence     93526   
2568  5709-LVOEQ      1  United States  California       San Mateo     94401   
2667  4367-NUYAO      1  United States  California       Cupertino     95014   
2856  1371-DWPAZ      1  United States  California        Redcrest     95569   
4331  7644-OMVMY      1  United States  California     Los Angeles     90029   
4687  3213-VVOLG      1  United States  California        Sun City     92585   
5104  2520-SGTTA      1  United States  California      Ben Lomond     95005   
5719  2923-ARZLG      1  United States  California        La Verne     91750   
6772  4075-WKNIU      1  United States  California            Bell     90201   
6840  2775-SEFEE      1  United States  California      Wilmington     90744   

                    Lat Long   Latitude   Longitude  Gender  ...  Contract  \
2234  34.084909, -117.258107  34.084909 -117.258107  Female  ...  Two year   
2438  36.869584, -118.189241  36.869584 -118.189241    Male  ...  Two year   
2568  37.590421, -122.306467  37.590421 -122.306467  Female  ...  Two year   
2667  37.306612, -122.080621  37.306612 -122.080621    Male  ...  Two year   
2856  40.363446, -123.835041  40.363446 -123.835041  Female  ...  Two year   
4331  34.089953, -118.294824  34.089953 -118.294824    Male  ...  Two year   
4687  33.739412, -117.173334  33.739412 -117.173334    Male  ...  Two year   
5104  37.078873, -122.090386  37.078873 -122.090386  Female  ...  Two year   
5719  34.144703, -117.770299  34.144703 -117.770299    Male  ...  One year   
6772  33.970343, -118.171368  33.970343 -118.171368  Female  ...  Two year   
6840  33.782068, -118.262263  33.782068 -118.262263    Male  ...  Two year   

     Paperless Billing             Payment Method  Monthly Charges  \
2234               Yes  Bank transfer (automatic)            52.55   
2438                No               Mailed check            20.25   
2568                No               Mailed check            80.85   
2667                No               Mailed check            25.75   
2856                No    Credit card (automatic)            56.05   
4331                No               Mailed check            19.85   
4687                No               Mailed check            25.35   
5104                No               Mailed check            20.00   
5719               Yes               Mailed check            19.70   
6772                No               Mailed check            73.35   
6840               Yes  Bank transfer (automatic)            61.90   

     Total Charges Churn Label Churn Value Churn Score  CLTV Churn Reason  
2234           NaN          No           0          36  2578          NaN  
2438           NaN          No           0          68  5504          NaN  
2568           NaN          No           0          45  2048          NaN  
2667           NaN          No           0          48  4950          NaN  
2856           NaN          No           0          30  4740          NaN  
4331           NaN          No           0          53  2019          NaN  
4687           NaN          No           0          49  2299          NaN  
5104           NaN          No           0          27  3763          NaN  
5719           NaN          No           0          69  4890          NaN  
6772           NaN          No           0          44  2342          NaN  
6840           NaN          No           0          65  5188          NaN  

[11 rows x 33 columns]

We see that all clients for which there is no 'Total Charges' value are not in the churn, and also almost all of them have the same contract type-'Two years', and one customer has 'One Year' contract

### Replacing missing values

Lets replace missing values with those calculated through monthly charges
We know how many months customer is using service, so we can check if calculation: monthly charges * tenure months will have result equal to total charges value

In [13]:
# Calculating charges
df['calc_charges'] = df['Monthly Charges'] * df['Tenure Months']

In [14]:
# calculating differnce between Total Charges and calculated charges
df['diff_in_charges'] = df['Total Charges'] - df['calc_charges']

In [15]:
# Let's check if our idea is working
fig = px.histogram(df, x="diff_in_charges", color = 'Contract', marginal="box")
fig.show()

Quantiles:

In [16]:
df.groupby('Contract')[['Total Charges', 'diff_in_charges']].quantile([.50, .80, .90, .95])

Total Charges  diff_in_charges
Contract                                           
Month-to-month 0.50       679.5500           0.0000
               0.80      2485.7300          24.8100
               0.90      3844.0600          54.0200
               0.95      4966.9200          85.3300
One year       0.50      2657.5500           0.7750
               0.80      5286.4600          55.0500
               0.90      6341.2500          92.2000
               0.95      7072.4725         133.3375
Two year       0.50      3623.9500           0.5000
               0.80      6399.2400          61.5300
               0.90      7457.6100          97.5700
               0.95      7922.3400         139.1800

For 95% of customers the difference is less than 10% of the median total charges. In this case, it seem possible to replace the missing values of total charges with those calculated in terms of monthly charges and not delete these rows.

In [17]:
df['Total Charges'] = np.where(df['Total Charges'].isna() == True,df['calc_charges'], df['Total Charges'])

In [18]:
df = df.drop(['calc_charges', 'diff_in_charges'], axis=1)

#### Exploratory Data Analysis
 
1. **General Churn Rate** 
The goal is to build a model that will predict customer churn. We need to firt understand what factors have the greatest impact on the churn of customers from the service. Churn customers are customers who have stopped using the service
Churn rate(the percentage of customers who churned), lets look at this metric in terms of the characteristics of the customers that we have.

In [19]:
fig = px.pie(df.groupby('Churn Label')['CustomerID'].nunique().reset_index(),
             values='CustomerID',
             names='Churn Label')
fig.show()

Among the customers in the dataset, 26.5% of customers are in churn and have stopped using the company's services

Understanding the reasons for the outflow of customers is a critically important task. The ability to predict that a customer may go into outflow can help save customers and profits

2. **Geography of the Customer**


In [20]:
# Let's start with the geography of customers
df.groupby(['Country', 'State'])['CustomerID'].count()

Country        State     
United States  California    7043
Name: CustomerID, dtype: int64

In [21]:
df['City'].nunique()

1129

There are 1129 unique cities in total, let's check where our customers are located and if geography has impact to the churn rate

In [23]:
fig = px.scatter_mapbox(df.groupby(['Latitude', 'Longitude'])['CustomerID'].count().reset_index(), lat="Latitude", lon="Latitude", hover_data= ['CustomerID'], zoom=4, height=300)
fig.update_layout(mapbox_style="open-street-map")
fig.update_layout(margin={"r":0, "t":0, "l":0, "b":0})
fig.show()